src/OrleansRuntime/Messaging/GatewayAcceptor.cs

```diff
@@ -41,7 +41,10 @@ protected override bool RecordOpenedSocket(Socket sock)
             {
                 //convert handshake cliendId to a GeoClient ID 
                 if (!string.IsNullOrEmpty(Silo.CurrentSilo.ClusterId))
+                {
                     client = GrainId.NewClientId(client.PrimaryKey, Silo.CurrentSilo.ClusterId);
+                    adjustReturnAddressOnNextMessage = true;
```

Comment:
Isn't there a race here? The global `adjustReturnAddressOnNextMessage` is set to `true` for each socket being accepted? Shouldn't the flag be tied to specific client/socket instead?